### DISCUSSION TITLE
Can't embed SWF in HTML?

### DISCUSSION THREAD

**Discussion by Fancy2209** [airsdk/Adobe-Runtime-Support#3232] [Q&A] [answered]
Hey, I was just testing AIR's ability to load a SWF from a HTML, but it says it's not supported?
![imagem](https://github.com/airsdk/Adobe-Runtime-Support/assets/64917206/68085520-b559-4505-851c-906a722ace81)
Is loading a swf through a object not enough anymore?

**Comment by ajwfrost** [comment, marked as answer]
Loading a SWF within HTML requires the Flash Player .. which would normally need to be installed separately on an end user's computer .. which isn't generally going to be possible now that Adobe have stopped supporting it.

We still support the Flash Player for enterprise customers though, so if you need this capability, please reach out to us separately (but the costs are significantly higher than for AIR SDK, it's meant to be a temporary solution for enterprise usage only - most of our customers are financial/retail/business services etc). If this is of interest, please contact pbsupport@harman.com and let them know where you're based, and they can hook you up with a regional sales director.

**Reply by Fancy2209** [reply]
Yeah I was just doing some tests, I cannot afford Enterprise stuff currently, thank you for the information though, I'll probably just use HTMLLoader instead of ExternalInterface.
Researching on this topic has made me curious though, do you know why the NPAPI plugin started being used? I've heard AIR21 didn't use the plugin, only 22+.

**Reply by ajwfrost** [reply]
As far as I'm aware, the use of SWFs within HTML within AIR has always been via the WebKit route i.e. either an HTML-based AIR application, or the use of HTMLLoader within an SWF-based AIR app. And WebKit's mechanism for browser plug-ins is NPAPI, the same as Firefox/Mozilla/Netscape (and Safari). Microsoft had the ActiveX-based plug-ins, and Google created the PPAPI-based plug-ins for Chromium, but when using AIR it's always been with the Adobe AIR-internal build of WebKit and NPAPI.

Adobe used to distribute the Flash Player along with AIR SDK .. but that meant the AIR SDK releases got tarred with the same security vulnerability reports as the Flash Player that it shipped with, so they switched to distributing a dummy/placeholder Flash Player binary, which (if there was no NPAPI Flash Player installed on the machine) gave users a link for downloading/installing it.

I forget which actual versions were involved in all this.. it's quite a while ago though!

FYI if you're trying to use HTMLLoader, then stick with AIR SDK 33.1.1.743 which has the WebKit capability still in it. Unless it's the kind of simpler usage that you could update to use StageWebView ....


thanks

**Reply by Fancy2209** [reply]
Good to know HTMLLoader doesn't work with WebView 2 on AIR.
I need HTMLLoader to replace ExternalInterface, that is used to log-in to discord to use it's API, so I can't really use StageWebView as I can't use AS and JS together without embeds. 
Still your information has been really helpful, thanks!
